'Set / Open Active Rule Set' runs into problems when ruleset file is a linked file

1. Add a rule set file as a linked file to a project
2. Right click the rule set file and click 'Set Active Rule Set' - the operation succeeds and there are no errors
3. Now click on References -> Analyzers -> 'Open Active Rule Set' for the project

**Observed:**
![capture](https://cloud.githubusercontent.com/assets/10579684/6835886/e6fdb520-d2fa-11e4-9fa7-0f26a27bbaa8.PNG)

Building the project from command line results in the following error -
C:\Program Files (x86)\MSBuild\14.0\bin\Microsoft.VisualBasic.CurrentVersion.targets(133,9): warning MSB3884: Could not find rule set file "MyRules.ruleset".

Looks like we add the following property in the project. We should probably include relative path to the linked file in such cases instead?
    <CodeAnalysisRuleSet>MyRules.ruleset</CodeAnalysisRuleSet>

**Expected:** Linked rule set file should be opened with the 'Project' context drop down set to the above project. If this is an unsupported scenario then the 'Set Active Rule Set' option should not be offered for linked files - I think we should support this - allowing linked rule set files would make it easy to share a single rule set file between multiple projects via inside VS.

(Note: If we decide to support this, please test the case where the rule set is added under a top level solution folder and linked in to multiple projects. Double clicking on the rule set from the solution folder should populate the Project drop down with all projects that use this rule set - with the 'active' context set to '&lt;All&gt;'. Double clicking on the rule set from a particular project where it is linked should set the 'active' context to that project.)

--- a/src/VisualStudio/Core/Def/Implementation/ProjectSystem/Extensions/IVsHierarchyExtensions.cs
+++ b/src/VisualStudio/Core/Def/Implementation/ProjectSystem/Extensions/IVsHierarchyExtensions.cs
@@ -63,6 +63,11 @@ namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
             return hierarchy.TryGetItemProperty(itemId, __VSHPROPID.VSHPROPID_Name, out name);
         }
 
+        public static bool TryGetCanonicalName(this IVsHierarchy hierarchy, uint itemId, out string name)
+        {
+            return ErrorHandler.Succeeded(hierarchy.GetCanonicalName(itemId, out name));
+        }
+
         public static bool TryGetParentHierarchy(this IVsHierarchy hierarchy, out IVsHierarchy parentHierarchy)
         {
             return hierarchy.TryGetProperty(__VSHPROPID.VSHPROPID_ParentHierarchy, out parentHierarchy);

--- a/src/VisualStudio/Core/SolutionExplorerShim/AnalyzersCommandHandler.cs
+++ b/src/VisualStudio/Core/SolutionExplorerShim/AnalyzersCommandHandler.cs
@@ -21,6 +21,7 @@ using Microsoft.VisualStudio.LanguageServices.Implementation.Utilities;
 using Microsoft.VisualStudio.LanguageServices.SolutionExplorer;
 using Microsoft.VisualStudio.Shell;
 using Microsoft.VisualStudio.Shell.Interop;
+using Roslyn.Utilities;
 using VSLangProj140;
 
 namespace Microsoft.VisualStudio.LanguageServices.Implementation.SolutionExplorer
@@ -485,11 +486,14 @@ namespace Microsoft.VisualStudio.LanguageServices.Implementation.SolutionExplore
         private void SetActiveRuleSetHandler(object sender, EventArgs e)
         {
             EnvDTE.Project project;
-            string fileName;
+            string ruleSetFileFullPath;
             if (_tracker.SelectedHierarchy.TryGetProject(out project) &&
-                _tracker.SelectedHierarchy.TryGetItemName(_tracker.SelectedItemId, out fileName))
+                _tracker.SelectedHierarchy.TryGetCanonicalName(_tracker.SelectedItemId, out ruleSetFileFullPath))
             {
-                UpdateProjectConfigurationsToUseRuleSetFile(project, fileName);
+                string projectDirectoryFullPath = Path.GetDirectoryName(project.FullName);
+                string ruleSetFileRelativePath = FilePathUtilities.GetRelativePath(projectDirectoryFullPath, ruleSetFileFullPath);
+
+                UpdateProjectConfigurationsToUseRuleSetFile(project, ruleSetFileRelativePath);
             }
         }
 

--- a/src/VisualStudio/Core/SolutionExplorerShim/SolutionExplorerShim.csproj
+++ b/src/VisualStudio/Core/SolutionExplorerShim/SolutionExplorerShim.csproj
@@ -34,6 +34,10 @@
       <Project>{1EE8CAD3-55F9-4D91-96B2-084641DA9A6C}</Project>
       <Name>CodeAnalysis</Name>
     </ProjectReference>
+    <ProjectReference Include="..\..\..\Workspaces\Core\Desktop\Workspaces.Desktop.csproj">
+      <Project>{2e87fa96-50bb-4607-8676-46521599f998}</Project>
+      <Name>Workspaces.Desktop</Name>
+    </ProjectReference>
     <ProjectReference Include="..\..\..\Workspaces\Core\Portable\Workspaces.csproj">
       <Project>{5F8D2414-064A-4B3A-9B42-8E2A04246BE5}</Project>
       <Name>Workspaces</Name>

--- a/src/Workspaces/Core/Desktop/Workspaces.Desktop.csproj
+++ b/src/Workspaces/Core/Desktop/Workspaces.Desktop.csproj
@@ -136,6 +136,7 @@
     <InternalsVisibleTo Include="Microsoft.VisualStudio.LanguageServices" />
     <InternalsVisibleTo Include="Microsoft.VisualStudio.LanguageServices.CSharp" />
     <InternalsVisibleTo Include="Microsoft.VisualStudio.LanguageServices.Implementation" />
+    <InternalsVisibleTo Include="Microsoft.VisualStudio.LanguageServices.SolutionExplorer" />
     <InternalsVisibleTo Include="Microsoft.VisualStudio.LanguageServices.Telemetry" />
     <InternalsVisibleTo Include="Microsoft.VisualStudio.LanguageServices.VisualBasic" />
     <InternalsVisibleTo Include="Roslyn.InteractiveFeatures" />

